locator/token_metadata.cc

```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
-        calculate_pending_ranges_for_leaving(unpimplified_this, strategy, new_pending_ranges, all_left_metadata);
-        // At this stage newPendingRanges has been updated according to leave operations. We can
-        // now continue the calculation by checking bootstrapping nodes.
-        calculate_pending_ranges_for_bootstrap(strategy, new_pending_ranges, all_left_metadata, get_dc_rack);
-        all_left_metadata->clear_gently().get();
-
-        // At this stage newPendingRanges has been updated according to leaving and bootstrapping nodes.
-        set_pending_ranges(keyspace_name, std::move(new_pending_ranges), can_yield::yes);
+        // new_token_metadata incorporates all the changes from leaving, bootstrapping and replacing
+        const auto new_token_metadata = token_metadata([&]() -> std::unique_ptr<token_metadata_impl> {
+            auto result = clone_only_token_map(false).get0();
+
+            // construct new_normal_tokens based on _bootstrap_tokens and _replacing_endpoints
+            std::unordered_map<inet_address, std::unordered_set<token>> new_normal_tokens;
+            if (!_replacing_endpoints.empty()) {
+                for (const auto& [token, inet_address]: _token_to_endpoint_map) {
+                    const auto it = _replacing_endpoints.find(inet_address);
+                    if (it == _replacing_endpoints.end()) {
+                        continue;
+                    }
+                    new_normal_tokens[it->second].insert(token);
+                }
+                for (const auto& [replace_from, replace_to]: _replacing_endpoints) {
+                    result->remove_endpoint(replace_from);
+                }
+            }
+            for (const auto& [token, inet_address]: _bootstrap_tokens) {
+                new_normal_tokens[inet_address].insert(token);
+            }
+            // apply new_normal_tokens
+            for (auto& [endpoint, tokens]: new_normal_tokens) {
+                result->update_topology(endpoint,
+                    get_dc_rack(endpoint),
+                    _normal_token_owners.contains(endpoint) ? node::state::normal : node::state::joining);
+                result->update_normal_tokens(std::move(tokens), endpoint).get();
+            }
+            // apply leaving endpoints
+            for (const auto& endpoint: _leaving_endpoints) {
+                result->remove_endpoint(endpoint);
+            }
+            result->sort_tokens();
+            return result;
+        }());
+
+        // merge tokens from token_to_endpoint and bootstrap_tokens,
+        // preserving tokens of leaving endpoints
+        const auto tokens = [&]() -> std::vector<dht::token> {
+            auto tokens = std::vector<dht::token>();
+            tokens.reserve(sorted_tokens().size() + get_bootstrap_tokens().size());
+            tokens.resize(sorted_tokens().size());
+            std::copy(begin(sorted_tokens()), end(sorted_tokens()), begin(tokens));
```

Comment:
why is `resize` needed?

typically, this is done as `std::copy(sorted_tokens().begin(), sorted_tokens().end(), std::back_inserter(tokens))`
does that prevent bulk copying (`dht::token` should have a trivial copy constructor, no?)

Reply:
```
-O3

resize -> memmove
back_inserter -> __copy_move<false, false, ...> -> copy item by item
```

https://godbolt.org/z/7Tz7E4Mqv

Reply:
Hmm, so this could probably be the most efficient way:
```c++
    auto tokens = sorted_tokens();
    tokens.reserve(tokens().size() + get_bootstrap_tokens().size());
    boost::range::copy(get_bootstrap_tokens() | boost::adaptors::map_keys, std::back_inserter(tokens));
```

Reply:
I don't like this boost magic, the variant with loop is simple and straightforward.

And it's difficult to decipher what's going on in the assembly https://godbolt.org/z/xa1TKnE6j. I doubt it's more efficient than the simple loop, in either way we need to collect the values from different memory locations/hashtable buckets.

Reply:
I do not insist of course. You are not the only one who dislikes boost (I do not mind it), but you do use `std::copy()` above instead of an open loop (which is good). Using an algorithm is better than open code a loop because it clearly shows the intention of the code without a need to read the loop. IIRC we do not use c++20 ranges because clang does not support it this is why we have to use boost sometimes.

Reply:
ok, this should be too frequently called anyway, so we don't need to squeeze every cpu cycle out of it.